typed-racket-lib/typed-racket/infer/restrict.rkt

```diff
@@ -34,35 +33,40 @@
       ;; structural recursion on types
       [((Pair: a1 d1) (Pair: a2 d2)) 
        (build-type -pair 
-                   (restrict* a1 a2 pref resolved) 
-                   (restrict* d1 d2 pref resolved))]
+                   (restrict* a1 a2 resolved) 
+                   (restrict* d1 d2 resolved))]
       ;; FIXME: support structural updating for structs when structs are updated to
       ;; contain not only *if* they are polymorphic, but *which* fields are too  
       ;;[((Struct: _ _ _ _ _ _)
       ;; (Struct: _ _ _ _ _ _))]
       [((Syntax: t1*) (Syntax: t2*))
-       (build-type -Syntax (restrict* t1* t2* pref resolved))]
+       (build-type -Syntax (restrict* t1* t2* resolved))]
       [((Promise: t1*) (Promise: t2*))
-       (build-type -Promise (restrict* t1* t2* pref resolved))]
+       (build-type -Promise (restrict* t1* t2* resolved))]
       
       ;; unions
-      [((Union: t1s) _) (apply Un (map (λ (t1*) (restrict* t1* t2 pref resolved)) t1s))]
-      [(_ (Union: t2s)) (apply Un (map (λ (t2*) (restrict* t1 t2* pref resolved)) t2s))]
+      [((Union: t1s) _) (apply Un (map (λ (t1*) (restrict* t1* t2 resolved)) t1s))]
+      [(_ (Union: t2s)) (apply Un (map (λ (t2*) (restrict* t1 t2* resolved)) t2s))]
+
+      ;; intersections
+      [((Intersection: t1s) _)
+       (apply -unsafe-intersect (for/list ([t1 (in-immutable-set t1s)])
+                                  (restrict* t1 t2 resolved)))]
+      [(_ (Intersection: t2s))
+       (apply -unsafe-intersect (for/list ([t2 (in-immutable-set t2s)])
+                                  (restrict* t1 t2 resolved)))]
       
       ;; resolve resolvable types if we haven't already done so
       [((? needs-resolving?) _) #:when (not (set-member? resolved (cons t1 t2)))
-                                (restrict* (resolve t1) t2 pref (set-add resolved (cons t1 t2)))]
+                                (restrict* (resolve t1) t2 (set-add resolved (cons t1 t2)))]
       [(_ (? needs-resolving?)) #:when (not (set-member? resolved (cons t1 t2)))
-                                (restrict* t1 (resolve t2) pref (set-add resolved (cons t1 t2)))]
+                                (restrict* t1 (resolve t2) (set-add resolved (cons t1 t2)))]
       
       ;; we don't actually want this - want something that's a part of t1
       [(_ _) #:when (subtype t2 t1)
              t2]
-      
-      ;; there's no overlap, so the restriction is empty
-      [(_ _) #:when (not (overlap t1 t2)) 
```

Comment:
Building an intersection will produce bot if there is no overlap, so this check is redundant.